parser: fix setting default value for dicts

MSusik opened the pull request:
The same as : https://github.com/inveniosoftware/invenio/pull/2749

Commit:
jsonalchemy: fix setting default value for dicts

Signed-off-by: Mateusz Susik <mateusz.susik@cern.ch>
Reviewed-by: Dimitrios Semitsoglou-Tsiapos <dsemitso@cern.ch>

jirikuncar commented:
@MSusik FYI it's waiting for @egabancho's review.

egabancho commented on jsonalchemy/validator.py:
```diff
@@ -113,9 +115,23 @@ def _validate(self, document, schema=None, update=False):
                 if not self.allow_unknown:
                     self._error(field, errors.ERROR_UNKNOWN_FIELD)
 
+        # Remove special rules, as they can cause error in Cerberus validation.
```
I think we should not do that, this way we are not applying this validation rules.
Actually the `_validate` method could be removed see https://github.com/inveniosoftware/jsonalchemy/pull/13

Maybe is worth to check what Cerberus gets from us and why it fails.
I think we shouldn't do this, we should instead check why Cerberus crashes

MSusik replied:
Cerberus gets from us whole schema, with special rules inside. That's why it fails.

After #13 was merged, this is obsolete anyways.

egabancho commented on jsonalchemy/reader.py:
```diff
@@ -422,8 +422,11 @@ def set_default_value(field, schema):
+                    old_value = self._json.__getitem__(
+                        field_name, exclude=['decorators', 'extensions'])
+                except KeyError:
+                    old_value = value
```
I don't fully understand what you are trying to do here. Maybe a test case would help ;-)

MSusik replied:
I wrote you some time ago on `gitter` what the error was about. As you couldn't explain to me what this code was supposed to do, I just caught the error that was produced by this line (I think that the fact that error is thrown is wrong), and secured it. IMO in the example given by me, the `if value is not None:` should not be passed, but it is. Unfortunately this is only a guess from me, I don't want to follow all the changes on the dictionary used in the module.

And I will add here my message from gitter

```
There is a function in reader.py called set_default_value
It creates a default value in case no other is provided.
In case of dictionary it recursively put Nones in every field
So, in this case, it returns a dictionary
Then, the _set_default_value encounters if value is not None, and the not provided value passes through this
Every other fields, that is not provided and is not a dictionary doesn't go there as set_default_value returns None
Then, in case of the dictionary, the wrapper doesn't have this key set in self._dict, which apparently happens only when the default value is provided.
```

egabancho replied:
I remember what you told me, but it would be nice if you also share it with the rest ;-)

And, indeed, an example or test case would be nice also.

jmartinm commented:
This PR seems to be blocking the inclusion of the author model in INSPIRE https://github.com/inspirehep/inspire-next/pull/170 is there something else to be done or should it be merged?

tiborsimko commented:
> is there something else to be done or should it be merged?

IIRC @egabancho asked for a test case, which I'd fully second. It's the best way to keep things working over longer term.   TDD rocks :smile:

MSusik commented:
Sorry, I was quite busy lately. I will create a testcase today in the afternoon.

Commit:
tests: test for the change in old_value

Signed-off-by: Mateusz Susik <mateusz.susik@cern.ch>

MSusik commented:
Added test.

jirikuncar commented:
@MSusik after discussion with @jalavik we would like to integrate the same patch for `invenio:maint-2.0` and `jsonalchemy` as it's not clear how quickly we progress with pythonic *dsl* implementation.

MSusik commented:
@jirikuncar 
As requested, I rebased in both places. Also, I removed one of the commits, which added a change to a code, which was already handled by @egabancho .
